### ISSUE TITLE
[Easyget Classic Arcade Game DIY] Arcade controls not responding on MakeCode

### ISSUE THREAD

**Issue by Lrigby22** [Vegz78/McAirpos#44]
Hi, I have just put together a Retropie arcade machine using custom Arcade controls (https://www.amazon.com/gp/product/B00WDQWK5M) and running through a Raspberry Pi 3.
I can successfully get NES and SNES games working properly, and after a while I managed to get MakeCode installed too - but whenever I open one of the elf. games the controls do not respond at all (not even to quit meaning a manual shutdown/restart each time).
I'm very new at this - but where should I look to fix the controls?
Any help would be appreciated as I'm an absolute beginner!
Thanks

**Comment by Vegz78** [comment]
Hi @Lrigby22,

Thanks for getting touch and trying MakeCode Arcade(MCA) games in RetroPie!

Hmmm, I do not think those Easyget Classic Arcade Game DIY controllers have been tested before with McAirpos, at least not as far as I know....

1. Did you install McAirpos using the automatic installation script or manual installation steps here?:
https://github.com/Vegz78/McAirpos#quick-automatic-installation-for-retropierpi-os

2. What does your /tmp/McAirpos.log says when attempting to start a MCA game with McAirpos? Please provide a copy of this file.

3. In general, the easiest way to get MCA games to work on a RPi with McAirpos, is to first connect the controller direction and button wires physically in accordance with any of the already working profiles included in McAirpos. This is because, since MCA games use lowest level controller signals - the raw Linux Input Subsystem -, it is less work to get your DIY controllers working here first, then easily reconfigure the controllers on a higher level inside RetroPie afterwards(RetroPie controller configurations have nothing to do with the controller configurations inside MCA games, but you can configure them the same).

Firstly, I would get acquainted with the `evtest` command line utility. Connect to the corresponding /dev/inputX device for your controllers in `evtest` and rewire so that `left` direction on stick gives something resembling `left` in evtest, same for other directions, then button A etc., etc. Then, when working in MCA games, reconfigure inside RetroPie accordingly.

Depending on your controllers, which may be EV_KEY(like a keyboard) instead of EV_ABS(analog sticks), it might be easier instead to use the `nomap` option with McAirpos' `launCharc` launcher, drop going via `uinput-mapper` and configure _/sd/arcade.cfg_ directly according to you controller directions and buttons results in `evtest`.

Please see here for some more detailed help:
https://github.com/Vegz78/McAirpos#default-and-modifying-the-layout-for-controls
https://github.com/Vegz78/McAirpos#Usage

And search the issues section for solutions for similar DYI controllers, like the EG STARTS and Zero Delay Encoder etc.:
https://github.com/Vegz78/McAirpos/issues?q=is%3Aissue+label%3AController+

**Comment by Vegz78** [comment]
Did you get this to work, @Lrigby22?

**Comment by Lrigby22** [comment]
Hi, sorry for not responding. I've just returned to work and haven't go around to trying yet! I'll be attempting it next week so I'll let you know how I get on.
Thanks

**Comment by edthedev** [comment]
I use MCA via McAirpos with a very similar custom arcade hardware setup (also using Zero Delay USB Encoders) and didn't have any issues getting the controls to respond.

One thing to check is whether the encoder has a 'mode' inuput.

Once when I did this I had to temporarily connect an input to the 'mode' spot on the USB Encoder, and (while powered up) hit the button the right number of times to put the encoder into 'digital', rather than 'analog' mode.

**Comment by Lrigby22** [comment]
Hi, sorry for the delay.
I did the automatic installation and the Pi works perfectly with NES/SNES etc.
I'm a complete newbie so I'm really not sure where to find Logs and encoder modes etc. Is there anyway you could give me a step by step on how to find these options? (Pictures and/or videos would be brilliant - but I appreciate time consuming!)
I think I will need to rewire them manually which I will attempt next, however when I managed to do an evtest, the code is running so fast I couldn't read it!
 Thanks again

**Comment by Vegz78** [comment]
Hi @Lrigby22,

Your ask is a little big, given that what you ask for, except videos, already exists in the main repository readme:
https://github.com/Vegz78/McAirpos/blob/master/README.md

Please take the time to read.

There are also many issues which resolves different controller hardware, and @edthedev just stepped in and gave you a major hint from his own experiences and contributions with the Zero Delay Encoder, and other similar DIY arcade controllers:
https://github.com/Vegz78/McAirpos/issues?q=is%3Aissue+"zero+delay"

This could all probably be better organized and written, but no one has the time, resources or motivation for doing this work, and I fail to see the revenue potential. You could always contribute yourself with the improvements you suggest, with a wiki here, own blogpost, entry to the forum.makecode.com, or whatever.


However, many in the community and I would be happy to help on specific points if you are stuck.


As noted earlier, MCA games operate directly on the lowest RAW Linux Input Subsystem for controllers(/dev/input/eventX).

You therefore have two options to achieve a working configuration:
-Either rewire your controller physically to correspond to its logical correct EV_KEY type keys/buttons by name from the start, or
-Configure either EV_KEY or EV_ABS type buttons in /SD/arcade.cfg together with the `nomap` option or /home/pi/McAirpos/McAirpos/uinput-mapper/configs/arcade1.py if the autoconfiguration in automatic mappings do not work.

The first option is by far the easiest path forward, but might require a controller mode change from EV_ABS to EV_KEY for the easiest operation, as @edthedev notes. Though, often, the DIY controllers are already configured as EV_KEY, like a keyboard.


`evtest` displays your current wiring setup, meaning the buttons configured for your controller as the first display before entering button presses. Then you can note down the configuration for each button press and do the physical rewiring. If too fast, break out for each button press, scroll up in the terminal and take notes and start over for each buttons.

Sometimes, a controller will continually output signals, even if not pressed(often EV_ABS controllers). You can then pipe `evtest` to a file, `evtest > bottons.txt`, break quickly after the button press, and search the file.


The log file, I have already provided you the path for earlier.


I know this is hard and frustrating when you have not done this before, but the main effort will be on your side, nevertheless.

Please read and tinker and try for a couple of days, and don't give up. Then come back here again if you are still stuck on a specific new thing. We will all also be glad if you get it to work and report back about that and how to do it, for others to read in the future.


Good luck!

**Comment by Vegz78** [comment]
Any progress with this, @Lrigby22?

**Comment by mrjshrimptonbis** [comment]
Vegz78, thank you so much for putting this together - I'm a teacher and I've got an arcade machine set up in my school which students helped build and now they are actually able to play their own games on it too, really cool!

I am unfortunately having the same issue as outlined above, the games launch and run just fine but all controller support for the entire system (Rasperry Pi B 3) cease to work once the game is launched.

I was hoping to be able to play the games initially via USB SNES controllers or the keyboard, but neither input methods work at all, the game launches and runs fine but there is no input detected. The above input methods work fine for all other emulation consoles on the machine. I'm running RetroPie on a Raspberry Pi 3B and I installed via the easy installation method outlined on your help page.

I'll do my best to go through the comments outlined above to fix this issue but if anyone else has any ideas how to resolve this issue please let me know 👍 

Again, regardless this is so cool and I appreciate your work so much Vegz78. The kids are thrilled to see their games on the machine!

**Comment by mrjshrimptonbis** [comment]
Further to the above... some things of note. I'm running RetroPie 4.8 on a Raspberry Pi 3B+. The games open and run absolutely great, it's just that all control support (including the keyboard) immediately ceases working once the game is open (making it also impossible to close the game).

Interestingly, the controls **sometimes** work for the first round of the game but as soon as the game restarts the system becomes unusable again. I've checked and trouble-shooted so much using the information shared above, my game controller USB SNES replica) and keyboard work perfectly in the main meno and also in the specified EV-test program outlined above. It seems to me that it's just selecting the wrong/a non-existing input device once the game starts playing. Everything functions perfectly on all other emulators on RetroPie.

![IMG_6072](https://github.com/user-attachments/assets/78f70842-6f3b-42a3-a09a-1dc780c78b20)
![IMG_6073](https://github.com/user-attachments/assets/52566c08-859a-43c1-a1be-4d9e51baa842)
![IMG_6074](https://github.com/user-attachments/assets/7d1fdac9-da6d-42da-9020-71574473fd1e)
![IMG_6077](https://github.com/user-attachments/assets/1453680e-b978-439c-9247-16be8c237402)
![IMG_6081](https://github.com/user-attachments/assets/e4ab100f-e0d8-4dbc-ad88-b60b811190ac)

**Comment by Vegz78** [comment]
Thanks for the nice feedback, @mrjshrimptonbis! I share your appreciation of seeing kids, my own or through teaching work, being able to run their own games on various arcades or handheld devices etc.! Also fun to see professional level adults on the MakeCode forum and here getting crazy advanced game projects running on the strangest of devices, and even to see some of my own much less interesting game creations run on some kind of hardware or game frontend, too. That is just the unique magic of MakeCode Arcade, and what got me into McAirpos in the first place (getting my son's MakeCode Arcade games to run on the Raspberry Pi).

Sorry for my late reply, but it has been really hard to follow up actively as the various use cases grew, breaking changes to the various retro gaming systems accelerated and as my son matured and lost interest and as life took twists and turns that lent little free time available for this. It has also been great to see that the community itself has taken over somewhat by sharing solutions and helping each other solve issues by themselves in my absence, lessening the sense of urgency of following up this project a great deal. -Thank you, everyone for your contributions!


However, I am planning to put a little effort now to get through some of the backlog now, whenever time allows.

So firstly, @mrjshrimptonbis and @Lrigby22: Were you able to solve the challenge in this issue eventually?

If so, I would greatly appreciate your feedback with the solution so I can close this issue with a solution that others in the same situation might benefit from.

If not, although I am a bit rusty myself on all the aspect of controller recognition and setup, I will try my best to offer some help:
1. From your provided pictures I get the impression that your controller is recognized as event0 by evtest and responds to button presses in the test loop.
2. However, as you describe, McAirpos somehow mixes up and recognizes event0(your gamepad) as keyboard and event3(your keyboard) as the gamepad.

This is probably due to the not 100% water tight recognition logic used in McAirpos for the uinput-mapper, where for all but a few special cases, USB keyboard is normally mapped before a game pad, but not always. If I remember correctly, I struggled with this myself in the beginning on one of my setups with a combined keyboard and mouse dongle, which prompted some changes in the recognition logic, which I thought would solve most cases. But obviously not every case...


To resolve this, i would firstly try 4 things:
1. Modify the launch command to use the keybswap argument, which was intended for exactly such normally rare cases where keyboard and gamepads are mixed up by the not bullet proof recognition script:
`~/McAirpos/McAirpos/launCharc/launCharc keybswap ~/RetroPie/roms/makecode/_gamefile.elf_`

2. Since you report that it sometimes works and sometimes not, it might indicate some changes made either by the user or the system in the USB connections to the system. A solution could be to stabilize what is causing this swap, if possible, or even better disconnect the keyboard altogether, leaving only the gamepad connected, which is then more easily recognized. (Access the device instead through ssh as not to confuse it with multiple competing input devices)

3. You could try to hardcode only the event number (event0) for the game pad in /sd/arcade.cfg and make this file write protected/read-only
 
4. As a last resort that is a bit more cumbersome, could be to hardcode /sd/arcade.cfg manually as described in the front page README.ME and modifying the launch command to use the nomap option (turn of automatic recognition and setup of controller):
`~/McAirpos/McAirpos/launCharc/launCharc nomap ~/RetroPie/roms/makecode/_gamefile.elf_`


I hope any of this could be of help and to hear back from you with a working solution!
